File: binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.nodes = []
    
    def inorderTraversalHelper(self, root: Optional[TreeNode]) -> List[int]:
        if root.left:
            self.inorderTraversalHelper(root.left)
        self.nodes.append(root.val)
        if root.right:
            self.inorderTraversalHelper(root.right)
    
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        self.inorderTraversalHelper(root)
        return self.nodes
    
## Can I do return only?
# class Solution:
#     def inorderTraversal(self, root: Optional[TreeNode], prev = []) -> List[int]:
#         if not root: return []
        
#         if root.left:
#             self.inorderTraversal(root.left, prev)
#         prev.append(root.val)
#         if root.right:
#             self.inorderTraversal(root.right, prev)
#         return prev

## Not sure why the single function version doesn't work

File: binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> ans_;
    Solution(){
    }
    
    void traverse(TreeNode* root){
        if (root->left){
            traverse(root->left);
        }
        ans_.push_back(root->val);
        if (root->right){
            traverse(root->right);
        }
    }
    
    vector<int> inorderTraversal(TreeNode* root) {
        if (!root){return {};}
        traverse(root);
        return ans_;
    }
};